docs/book/src/developer/testing.md

```diff
@@ -96,6 +142,62 @@ should take care of ensuring each test run in isolation from the others, by:
 - Avoiding object name conflict.
 - Cleaning up objects created by each test.
 
+This is an example of a test implementing those recommendations:
+
+```golang
+func TestAFunc(t *testing.T) {
+	g := NewWithT(t)
+	// Generate namespace with a random name starting with ns1; such namespace
+	// will host test objects in isolation from other tests.
+	ns1, err := testEnv.CreateNamespace(ctx, "ns1") 
+	g.Expect(err).ToNot(HaveOccurred())
+	defer func() {
+		// Cleanup the test namespace
+		g.Expect(testEnv.DeleteNamespace(ctx, ns1)).To(Succeed())
+	}()
+
+	obj := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test",
+			Namespace: ns1.Name, // Place test objects in the test namespace
+		},
+	}
+	
+	// Actual test code...
+}
+```
+
+In case of object used in many test case withing the same test, it is possible to leverage on Kuberentes `GenerateName`:
+
+```golang
+func TestAFunc(t *testing.T) {
+	// create test namespace
+
+	obj := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-",  // Instead of assigning a name, use GenerateName
+			Namespace:    ns1.Name, // Place test objects in the test namespace
+		},
+	}
+	
+	t.Run("test case 1", func(t *testing.T) {
+		g := NewWithT(t)
+		// Deep copy the object in each test case, so we prevent side effects in case the object changes.
+		// Additionally, thanks to GenerateName, the objects gets a new name for each test case.
+		obj := obj.DeepCopy()
+
+	    // Actual test case code...
+	}
+	t.Run("test case 2", func(t *testing.T) {
+		g := NewWithT(t)
+		obj := obj.DeepCopy()
+
+	    // Actual test case code...
+	}
+	// More test cases.
+}
+```
```

Comment:
I think we can combine the two examples into one.
```suggestion
This is an example of a test implementing those recommendations.
We can leverage the `GenerateName` property to generate unique names for our objects.
For objects that are shared across sub-tests, ensure they are scoped within the test namespace and deep copied to avoid cross-test changes that may occur to the object.

```golang
func TestAFunc(t *testing.T) {
	g := NewWithT(t)
	// Generate namespace with a random name starting with ns1; such namespace
	// will host test objects in isolation from other tests.
	ns1, err := testEnv.CreateNamespace(ctx, "ns1") 
	g.Expect(err).ToNot(HaveOccurred())
	defer func() {
		// Cleanup the test namespace
		g.Expect(testEnv.DeleteNamespace(ctx, ns1)).To(Succeed())
	}()

	obj := &clusterv1.Cluster{
		ObjectMeta: metav1.ObjectMeta{
			GenerateName: "test-",  // Instead of assigning a name, use GenerateName
			Namespace: ns1.Name, // Place test objects in the test namespace
		},
	}
		
	t.Run("test case 1", func(t *testing.T) {
		g := NewWithT(t)
		// Deep copy the object in each test case, so we prevent side effects in case the object changes.
		// Additionally, thanks to GenerateName, the objects gets a new name for each test case.
		obj := obj.DeepCopy()

	        // Actual test case code...
	}
	t.Run("test case 2", func(t *testing.T) {
		g := NewWithT(t)
		obj := obj.DeepCopy()

	        // Actual test case code...
	}
	// More test cases.
}
```
```

Reply:
I prefer to keep the two examples separated so we have both an example for simple test and the example for table-driven tests. Never the less I took your sentence about deep copying objects and included in the doc